### DISCUSSION TITLE
How to find out which own dependencies Webpack uses at runtime?

### DISCUSSION THREAD

**Discussion by mike-lischke** [webpack/webpack#12518] [General]
This is a pretty special question I have which [Stackoverflow](https://stackoverflow.com/questions/65920043/how-to-find-the-dependencies-of-the-webpack-runtime) is not able to answer.

What I need to know is what parts of Webpack and which 3rd party code it relies on are actually shipped in generated code. For legal approval I need to list all 3rd/4th party libraries that are shipped to customers.

I looked at the package.json file of the project, but to me it looks as if that contains many build/dev dependencies. I installed webpack@4.44.2 explicitly (this is the version used by current react-scripts) which leaves me with a 600 lines long dependency tree. This cannot all be runtime dependencies. Much of that is certainly for build/dev time only, right?

So, how can I tell which dep is shipped?

**Comment by alexander-akait** [comment]
What is wrong with tool that I recommend? What do you want from webpack here?

> This cannot all be runtime dependencies.

All dependencies are used, because they are in `dependencies`

**Reply by mike-lischke** [reply]
Hey Alex, well, I haven't come around to test it, but you stroke through your recommendation, leaving me in the belief you think that's a wrong answer. If that's not the case then take my apologies.

**Reply by alexander-akait** [reply]
Feel free to feedback

**Reply by mike-lischke** [reply]
https://github.com/webpack-contrib/webpack-bundle-analyzer is an awesome tool. Lots of informations and a visual tool to get a feeling for bundle sizes. Unfortunately, it doesn't answer my question about the 3rd party modules used at runtime from webpack.

**Reply by alexander-akait** [reply]
what do you mean `used at runtime from webpack`?

**Reply by mike-lischke** [reply]
You are the second one today asking what "runtime" or "used at runtime" means - that leaves me a bit perplexed.

The generated JS code that's executed when you open your web app in a browser is executed while the application runs, hence that's the `runtime`. The opposite is when you write your code and build it, which is commonly refered to as "build time", "development" or "design time".

Additionally the term `runtime` is also used for libraries that provides all the base functionality for your app (JS itself could be considered to be a runtime lib). A C++ developer would consider the STL as the runtime, or a C developer sees glibc as that.

Similarly, there must be something that Webpack uses to provide the functionality (e.g. for asynchronous loading of imports, in Webpack terms these are also dependencies). This code must come from somewhere, even if it is not loaded as separate module, but embedded into the final bundles. There's probably more from Webpack in the final bundles, but I'm pretty sure nothing from `acorn` is used there (to give an example), even though `acorn` is in Webpack's dependency list.

**Reply by mike-lischke** [reply]
Btw. even Babel uses the term `runtime`: https://www.npmjs.com/package/@babel/runtime

**Reply by alexander-akait** [reply]
`@babel/runtime` is not the same `The generated JS code that's executed`, you need plugin for this, what you need? Licenses or just list or licenses + list

**Reply by mike-lischke** [reply]
I would need a list of libraries used by webpack, that are shipped in the final bundles. I know that these libraries are not directly shipped, but embedded in the final code. Yet this code is licensed and I need to know from which library any code ended up in these final bundles.

**Reply by alexander-akait** [reply]
You can try https://github.com/xz64/license-webpack-plugin

**Reply by mike-lischke** [reply]
Thanks for your help, Alex. I already use that plugin for the other licenses. Only Webpack wasn't clear. However, I inspected more deeply what the bundle analyzer provides and I can see all the code and their node module origin. This is a super helpful analysis and this tool will certainly be part of my toolset for future tasks. In the graphic I see no reference to webpack, so I assume nothing from the webpack code actually ends up in the generated bundles. Webpack is a build-time only tool. Again, many thanks for your patience and help Alex!

**Reply by alexander-akait** [reply]
Yes, webpack doesn't include something extra, only own core

**Comment by sokra** [comment]
The webpack core doesn't add any 3rd party code automatically (plugins and other tooling might do).

But we add a bunch of runtime code depending on what you use. You find this code in this repo in these files:

- webpack/hot/*
- *.runtime.js
- *RuntimeModule.js

**Reply by mike-lischke** [reply]
Excellent, that's the answer I was hoping for. Many thanks @sokra.
